src/rpc_ctx.c

```diff
@@ -74,139 +74,137 @@ rpc_msg_init(struct rpc_msg *msg)
 	msg->RPCM_ack.ar_results.proc = (xdrproc_t) xdr_void;
 }
 
+/*
+ * On success, returns with RPC_CTX_FLAG_LOCKED followed by RPC_DPLX_FLAG_LOCKED
+ */
 rpc_ctx_t *
 rpc_ctx_alloc(CLIENT *clnt, rpcproc_t proc, xdrproc_t xdr_args, void *args_ptr,
 	      xdrproc_t xdr_results, void *results_ptr, struct timeval timeout)
 {
-	struct svc_vc_xprt *xd = (struct svc_vc_xprt *)clnt->cl_p1;
-	struct rpc_dplx_rec *rec = xd->rec;
+	struct cx_data *cx = CX_DATA(clnt);
+	struct rpc_dplx_rec *rec = cx->cx_rec;
+	struct svc_vc_xprt *xd = VC_DR(rec);
 	rpc_ctx_t *ctx = mem_alloc(sizeof(rpc_ctx_t));
 
 	rpc_msg_init(&ctx->cc_msg);
 
 	/* protects this */
 	mutex_init(&ctx->we.mtx, NULL);
+	mutex_lock(&ctx->we.mtx);
 	cond_init(&ctx->we.cv, 0, NULL);
+	ctx->flags = RPC_CTX_FLAG_NONE;
+	ctx->refcount = 1;
 
 	/* some of this looks like overkill;  it's here to support future,
 	 * fully async calls */
 	ctx->ctx_u.clnt.clnt = clnt;
 	ctx->ctx_u.clnt.timeout.tv_sec = 0;
 	ctx->ctx_u.clnt.timeout.tv_nsec = 0;
 	timespec_addms(&ctx->ctx_u.clnt.timeout, tv_to_ms(&timeout));
-	ctx->flags = 0;
 
 	/* this lock protects both xid and rbtree */
-	REC_LOCK(rec);
+	rpc_dplx_rli(rec);
 	ctx->xid = ++(xd->cx.calls.xid);
 
 	if (opr_rbtree_insert(&xd->cx.calls.t, &ctx->node_k)) {
-		REC_UNLOCK(rec);
 		__warnx(TIRPC_DEBUG_FLAG_ERROR,
 			"%s: call ctx insert failed (xid %d client %p)",
 			__func__, ctx->xid, clnt);
-		mutex_destroy(&ctx->we.mtx);
-		cond_destroy(&ctx->we.cv);
-		mem_free(ctx, sizeof(*ctx));
+		rpc_ctx_release(ctx);
+		rpc_dplx_rui(rec);
 		return (NULL);
 	}
-	REC_UNLOCK(rec);
 	return (ctx);
 }
 
-void
-rpc_ctx_next_xid(rpc_ctx_t *ctx, uint32_t flags)
+/* RPC_CTX_FLAG_LOCKED, RPC_DPLX_FLAG_LOCKED
+ */
+bool
+rpc_ctx_next_xid(rpc_ctx_t *ctx)
 {
 	CLIENT *clnt = ctx->ctx_u.clnt.clnt;
-	struct svc_vc_xprt *xd = (struct svc_vc_xprt *)clnt->cl_p1;
-	struct rpc_dplx_rec *rec = xd->rec;
+	struct cx_data *cx = CX_DATA(clnt);
+	struct rpc_dplx_rec *rec = cx->cx_rec;
+	struct svc_vc_xprt *xd = VC_DR(rec);
 
-	/* this lock protects both xid and rbtree */
-	REC_LOCK(rec);
+	/* the lock protects both xid and rbtree */
```

Comment:
Changed this comment to "the" lock, and listed the required locks at the top of the function.